// node-datachannel/dist/cjs/polyfill/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var RTCCertificate = require('./RTCCertificate.cjs');
var RTCDataChannel = require('./RTCDataChannel.cjs');
var RTCDtlsTransport = require('./RTCDtlsTransport.cjs');
var RTCIceCandidate = require('./RTCIceCandidate.cjs');
var RTCIceTransport = require('./RTCIceTransport.cjs');
var RTCPeerConnection = require('./RTCPeerConnection.cjs');
var RTCSctpTransport = require('./RTCSctpTransport.cjs');
var RTCSessionDescription = require('./RTCSessionDescription.cjs');
var Events = require('./Events.cjs');
var RTCError = require('./RTCError.cjs');

var p = {
  RTCCertificate: RTCCertificate.default,
  RTCDataChannel: RTCDataChannel.default,
  RTCDtlsTransport: RTCDtlsTransport.default,
  RTCIceCandidate: RTCIceCandidate.default,
  RTCIceTransport: RTCIceTransport.default,
  RTCPeerConnection: RTCPeerConnection.default,
  RTCSctpTransport: RTCSctpTransport.default,
  RTCSessionDescription: RTCSessionDescription.default,
  RTCDataChannelEvent: Events.RTCDataChannelEvent,
  RTCPeerConnectionIceEvent: Events.RTCPeerConnectionIceEvent,
  RTCError: RTCError.default
};

exports.RTCCertificate = RTCCertificate.default;
exports.RTCDataChannel = RTCDataChannel.default;
exports.RTCDtlsTransport = RTCDtlsTransport.default;
exports.RTCIceCandidate = RTCIceCandidate.default;
exports.RTCIceTransport = RTCIceTransport.default;
exports.RTCPeerConnection = RTCPeerConnection.default;
exports.RTCSctpTransport = RTCSctpTransport.default;
exports.RTCSessionDescription = RTCSessionDescription.default;
exports.RTCDataChannelEvent = Events.RTCDataChannelEvent;
exports.RTCPeerConnectionIceEvent = Events.RTCPeerConnectionIceEvent;
exports.RTCError = RTCError.default;
exports.default = p;
